### PR TITLE
[GSoC] Simplification for the solutions of systems of ODEs

### PR DESCRIPTION
**Pull request by mijo2**
#### Release Notes

* solvers
  * Added simplification strategies for simplifying the solutions of systems of ODEs

### CODE REVIEW

File: sympy/solvers/ode/systems.py [hunk 1]
```diff
@@ -1508,6 +1508,8 @@ def _higher_order_ode_solver(match):
     is_euler = is_transformed and type == "type1"
     is_higher_order_type2 = is_transformed and type == "type2" and 'P' in match
 
+    print(match.get('rhs', None))
```
Comment (czgdp1807) [inline sympy/solvers/ode/systems.py:1511]: I hope this was for debugging(similarly other such statements)?

Comment (mijo2) [reply sympy/solvers/ode/systems.py:1511]: Yes, I have removed these lines locally. Forgot to fix up these statements

**Comment by mijo2** [comment]
I have created a set of functions that can be used to print the test case given a system. Along with that, I have also created a basic structure to replace any rationals in a solution with `Rational(a, b)`. I am not sure if it will always work but the below is what I have created:

```
In [113]: def filldedent_string_to_tab(string): 
     ...:     string = filldedent(string, w=100) 
     ...:     string = string.split("\n") 
     ...:     if len(string[0]) == 0: del string[0] 
     ...:     string = ['  ' + s if i != 0 else s for i, s in enumerate(string)] 
     ...:     return "\n".join(string)

In [114]: def add_more_spaces(string, more_spaces=10, ignore_first=False): 
     ...:     string = string.split("\n") 
     ...:     string = [more_spaces * ' ' + s if i != 0 or not ignore_first else s for i, s in enumerate(string)] 
     ...:     return "\n".join(string) 
     ...:                             

In [114]: def replace_rationals(str):  
     ...:     array = re.findall(r'[0-9]+/[0-9]+', str)  
     ...:     for expr in array: 
     ...:         str = str.replace(expr, "Rational({}, {})".format(*expr.split('/'))) 
     ...:     return str 
     ...:                                                                                                                                                                                                   

In [115]: def print_eqs(sol, initial_string="sol = {}"): 
     ...:     base = len(initial_string) 
     ...:     print(initial_string, end="") 
     ...:     for i, eq in enumerate(sol): 
     ...:         print((i==0)*'[' + replace_rationals(add_more_spaces(filldedent_string_to_tab(str(eq)), more_spaces=base, ignore_first=(i == 0))) + (i == len(sol)-1) * ']') 
     ...:                           

In [116]: def print_one_sol(eqs, n=1): 
     ...:     sol = dsolve(eqs) 
     ...:     print_eqs(eqs, "eqs{} = ".format(n)) 
     ...:     print_eqs(sol, "sol{} = ".format(n)) 
     ...:     print("assert dsolve(eqs{}) == sol{}".format(n, n)) 
     ...:     print("assert checksysodesol(eqs{}, sol{}) == (True, {})".format(n, n, n*[0])) 
     ...:    
                                                                                                                                                                                                                                                                                                                                                                       
```

**Comment by oscarbenjamin** [comment]
@mijo2 what's the status of this?

Did you cancel the tests yourself?

**Comment by mijo2** [comment]
@oscarbenjamin I cancelled the tests. I want to complete this PR, but due to some commitments, I won't be able to work for atleast the first half of this month. I do intend to complete this PR in this month itself

**Comment by oscarbenjamin** [comment]
I've pushed a few commits on here to finish off the simplification. This involved changing many of the tests because the output from dsolve changed. Some tests were too slow so I've reorganised some to mark some as slow or perhaps too slow.

@mijo2 Can you review this?

**Comment by mijo2** [comment]
> I've pushed a few commits on here to finish off the simplification. This involved changing many of the tests because the output from dsolve changed. Some tests were too slow so I've reorganised some to mark some as slow or perhaps too slow.
> 
> @mijo2 Can you review this?

Sure, I review it soon, at max by tomorrow. Thanks for the commits.

**Comment by mijo2** [comment]
Looks good to me. Really liked how you implemented the simplification part along with optimizing the integral computations.

**Comment by oscarbenjamin** [comment]
Thanks @mijo2. I think with this in the new systems code is ready for release. It would be great if you could write a new page in the docs to describe the systems that can be solved and the new functions though. At the moment none of this is really visible to users except that `dsolve` might give better answers for many systems.

**Comment by mijo2** [comment]
> It would be great if you could write a new page in the docs to describe the systems that can be solved and the new functions though.

I would love to write a page showing all the examples that can be solved but I am currently very busy. Can you tell me a tentative date for the release so that I can make proper arrangements and plan accordingly to write the doc page?

**Comment by oscarbenjamin** [comment]
This PR was the main blocker for the release so now it should be fairly soon although I haven't announced anything yet.

**Pull request merged**